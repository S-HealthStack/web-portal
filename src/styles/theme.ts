import { css, DefaultTheme } from 'styled-components';
import { px } from './utils';

const colors = {
  // gray gradations
  black: '#000000',
  black40: 'rgba(0, 0, 0, 0.4)',
  black08: 'rgba(0, 0, 0, 0.08)',

  // light mode
  surface: '#FFFFFF',
  background: '#F7F8FA',
  onPrimary: '#FFFFFF',
  backgroundOnPrimary: '#FFFFFF',
  backgroundSurface: '#FFFFFF',
  backgroundOnSurface: '#2D2D2D',
  primary: '#4475E3',
  primaryLight: '#ECF1FC',
  primaryHovered: '#2F65E0',
  primary05: '#F6F8FE',
  primary10: '#ECF1FC',
  primary30: '#C7D6F7',
  primaryDisabled: '#B3C6F1',
  primaryLightPressed: '#E3EAFB',
  primaryLightFocused: '#C7D6F7',
  textPrimary: '#474747',
  primaryWhite: '#FFFFFF',
  disabled: '#DBDBDB',
  disabled20: 'rgba(219, 219, 219, 0.2)',
  onDisabled: '#A1A1A1',

  textPrimaryBlue: '#4475E3',
  textPrimaryDark: '#474747',
  textSecondaryGray: '#808080',
  textDisabled: '#838383',
  onSurface: '#2D2D2D',
  secondarySkyblue: '#00B0D7', // replace
  backgoundLight: '#F0F0F0',
  primaryBlueHovered: '#2F65E0',
  primaryBluePressed: '#2057D5',

  statusSuccess: '#2DC008',
  statusSuccess10: '#DCF2EA',
  statusSuccessText: '#00825D',

  statusError: '#FF3333',
  statusError10: '#FFEBEB',
  statusErrorText: '#D14343',

  statusWarning: '#F5BF00',
  statusWarning10: '#FFF7D0',
  statusWarningText: '#A36400',

  secondaryViolet: '#944ED7',
  secondaryVioletHovered: '#6A3799',
  secondaryViolet10: '#F4EDFB',

  secondaryTangerine: '#F39C18',
  secondaryTangerineHovered: '#CC8314',
  secondaryTangerine10: '#FEF5E8',

  secondarySkyBlue: '#00B0D7',
  secondarySkyBlueHovered: '#0092B3',
  secondarySkyBlue10: '#E5F7FB',

  secondaryGreen: '#1AD598',
  secondaryGreenHovered: '#16B37F',
  secondaryGreen10: '#E8FBF5',

  secondaryRed: '#FA5F4F',
  secondaryRedHovered: '#CC4E40',
  secondaryRed10: '#FFEFED',
};

export type SpecColorType = keyof typeof colors;

const weightToCssValue = {
  regular: 400,
  upperregular: 450,
  medium: 500,
  semibold: 600,
  bold: 700,
};

const fontSpec = (
  size: number,
  lineHeight: number | string,
  weight: keyof typeof weightToCssValue,
  fontFamily: string
) => css`
  font-family: ${fontFamily}, sans-serif;
  font-size: ${px(size)};
  line-height: ${typeof lineHeight === 'string' ? lineHeight : px(lineHeight)};
  font-weight: ${weightToCssValue[weight]};
  font-style: normal;
`;

const typography = {
  // Inter typeface
  labelRegular: fontSpec(10, 15, 'regular', 'Inter'),
  labelSemibold: fontSpec(10, 15, 'semibold', 'Inter'),

  bodyXSmallRegular: fontSpec(12, 18, 'regular', 'Inter'),
  bodyXSmallSemibold: fontSpec(12, 18, 'semibold', 'Inter'),
  bodySmallRegular: fontSpec(14, 21, 'regular', 'Inter'),
  bodySmallSemibold: fontSpec(14, 21, 'semibold', 'Inter'),
  bodyMediumRegular: fontSpec(16, 24, 'regular', 'Inter'),
  bodyMediumSemibold: fontSpec(16, 24, 'semibold', 'Inter'),
  bodyLargeRegular: fontSpec(18, 26, 'regular', 'Inter'),
  bodyLargeSemibold: fontSpec(18, 26, 'semibold', 'Inter'),

  headingXSmall: fontSpec(16, 21, 'semibold', 'Inter'),
  headingSmall: fontSpec(18, 23, 'semibold', 'Inter'),
  headingXMedium: fontSpec(18, 27, 'semibold', 'Inter'),
  headingXMediumRegular: fontSpec(18, 27, 'regular', 'Inter'),
  headingMedium: fontSpec(24, 32.2, 'semibold', 'Inter'),
  headingLargeSemibold: fontSpec(32, 42, 'semibold', 'Inter'),
  headingXLargeSemibold: fontSpec(48, 58, 'semibold', 'Inter'),
  headingXXLargeSemibold: fontSpec(60, 72, 'semibold', 'Inter'),

  // Fira Code typeface
  query14: fontSpec(14, '130%', 'upperregular', 'Fira Code'),
  query12: fontSpec(12, '150%', 'upperregular', 'Fira Code'),
};

const boxShadow = {
  card: `0 0 ${px(2)} ${px(0)} rgba(0,0,0,0.15)`,
  avatar: `0 ${px(8)} ${px(10)} ${px(2)} 0 rgba(0,0,0,0.12)`,
  studyAvatar: `0 ${px(15)} ${px(20)} rgba(180,180,180,0.3)`,
  previewScreen: `0 ${px(2)} ${px(4)} rgba(71,71,71,0.25)`,
};

const animation = {
  defaultTiming: 'cubic-bezier(0.45, 0.05, 0, 1)',
  rippleTiming: 'cubic-bezier(0.4, 0, 0.2, 1)',
};

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: typeof colors;
    typography: typeof typography;
    boxShadow: typeof boxShadow;
    animation: typeof animation;
  }
}

const theme: DefaultTheme = {
  colors,
  typography,
  boxShadow,
  animation,
};

export default theme;
