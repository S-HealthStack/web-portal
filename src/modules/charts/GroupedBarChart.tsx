import React, { useRef, useEffect, useMemo, useCallback } from 'react';
import { useUpdateEffect } from 'react-use';
import styled, { useTheme, css } from 'styled-components';
import * as d3 from 'd3';
import _upperFirst from 'lodash/upperFirst';

import { colors, typography } from 'src/styles';
import { SpecColorType } from 'src/styles/theme';
import { XAxis, YAxis } from './common-components';

const axisCommon = css`
  stroke-opacity: 1;
  stroke: ${colors.backgoundLight};
`;

const StyledSvg = styled.svg<{ isHorizontal: boolean }>`
  .xAxis {
    .tick {
      line {
        ${axisCommon}
      }
      ${({ isHorizontal }) =>
        isHorizontal
          ? css`
              &:first-child,
              &:last-child {
                line {
                  ${axisCommon}
                }
              }
              line {
                stroke-opacity: 0;
              }
            `
          : css`
              &:last-child {
                line {
                  stroke-dasharray: 1;
                }
              }
              line {
                stroke-dasharray: 2;
              }
            `}
      text {
        ${typography.labelRegular};
        color: ${colors.textPrimary};
      }
    }
  }

  .yAxis {
    .tick {
      line {
        ${axisCommon}
      }
      ${({ isHorizontal }) =>
        !isHorizontal &&
        css`
          &:first-child,
          &:last-child {
            line {
              stroke-dasharray: 1;
            }
          }
          line {
            stroke-dasharray: 2;
          }
        `}
      text {
        ${typography.labelRegular};
        color: ${colors.textPrimary};
      }
    }
  }
`;

type DataItem = {
  name: string;
  dataKey: string;
  value: number;
};

export type GroupedBarChartProps = {
  data: DataItem[];
  width: number;
  height: number;
  barColors: SpecColorType[];
  numberOfKeys: number;
  maxValue: number;
  isHorizontal?: boolean;
  formatXAxisTick?: (d: d3.NumberValue) => string;
  showYNegativeArea?: boolean;
  yTickValues?: Iterable<d3.NumberValue>;
  chartMargin?: { top: number; right: number; bottom: number; left: number };
};

const MARGIN = { top: 0, right: 12, bottom: 36, left: 53 };

const GroupedBarChart = ({
  width,
  height,
  data,
  barColors,
  numberOfKeys,
  maxValue,
  formatXAxisTick,
  isHorizontal,
  showYNegativeArea,
  yTickValues,
  chartMargin = MARGIN,
}: GroupedBarChartProps) => {
  const svgRef = useRef(null);
  const theme = useTheme();

  const x = useMemo(
    () => d3.map(data, (d) => (isHorizontal ? +d.value : d.dataKey)),
    [data, isHorizontal]
  );
  const y = useMemo(
    () => d3.map(data, (d) => (isHorizontal ? d.dataKey : +d.value)),
    [data, isHorizontal]
  );
  const z = useMemo(() => d3.map(data, (d) => d.name), [data]);

  const xRange = useMemo(() => [chartMargin.left, width - chartMargin.right], [chartMargin, width]);
  const yRange = useMemo(
    () => [height - chartMargin.bottom, chartMargin.top],
    [chartMargin, height]
  );

  const xScaleLinear = useMemo(
    () => d3.scaleLinear([0, maxValue], xRange).nice(),
    [maxValue, xRange]
  );
  const xScaleBand = useMemo(() => d3.scaleBand(x, xRange), [x, xRange]);

  const yScaleBand = useMemo(() => d3.scaleBand(y, yRange), [y, yRange]);

  const yScaleLinear = useMemo(
    () => d3.scaleLinear([showYNegativeArea ? -maxValue : 0, maxValue], yRange).nice(),
    [maxValue, showYNegativeArea, yRange]
  );

  const zScale = useMemo(
    () =>
      d3.scaleOrdinal(
        z,
        barColors.map((color) => theme.colors[color])
      ),
    [barColors, theme.colors, z]
  );
  const crossScale = useMemo(
    () =>
      d3
        .scaleBand(z, [0, (isHorizontal ? yScaleBand : xScaleBand).bandwidth()])
        .padding((2 / numberOfKeys) * 0.56),
    [isHorizontal, numberOfKeys, xScaleBand, yScaleBand, z]
  );

  const updateAxis = useCallback(() => {
    d3.select(svgRef.current).select('.upperChartBorder').remove();
    d3.select(svgRef.current).select('.rightChartBorder').remove();

    if (isHorizontal) {
      /* upper chart border */
      d3.select(svgRef.current)
        .append('line')
        .attr('class', 'upperChartBorder')
        .style('stroke', theme.colors.onSurface)
        .style('stroke-width', 1)
        .style('stroke-opacity', 0.2)
        .attr('x1', chartMargin.left)
        .attr('y1', 0)
        .attr('x2', width - chartMargin.right)
        .attr('y2', 0);
    } else {
      /* right chart border */
      d3.select(svgRef.current)
        .append('line')
        .attr('class', 'rightChartBorder')
        .style('stroke', theme.colors.onSurface)
        .style('stroke-width', 1)
        .style('stroke-dasharray', 1)
        .style('stroke-opacity', 0.2)
        .attr('x1', width - chartMargin.right)
        .attr('y1', 0)
        .attr('x2', width - chartMargin.right)
        .attr('y2', height - chartMargin.bottom);
    }
  }, [height, isHorizontal, theme.colors.onSurface, width, chartMargin]);

  const updateChart = useCallback(() => {
    updateAxis();

    d3.select(svgRef.current).select('.bars').selectAll('rect').remove();

    const xCoord = (i: number) =>
      isHorizontal ? xScaleLinear(0) : (xScaleBand(x[i] as number) || 0) + (crossScale(z[i]) || 0);

    const yCoord = (i: number) =>
      isHorizontal
        ? (yScaleBand(y[i] as number) || 0) + (crossScale(z[i]) || 0)
        : yScaleLinear(y[i] < 0 ? 0 : (y[i] as number));

    const barWidth = (i: number) =>
      isHorizontal ? xScaleLinear(x[i] as number) - xScaleLinear(0) : crossScale.bandwidth();

    const barHeight = (i: number) =>
      isHorizontal
        ? crossScale.bandwidth()
        : Math.abs(yScaleLinear(0) - yScaleLinear(y[i] as number));

    d3.select(svgRef.current)
      .select('.bars')
      .selectAll('rect')
      .data(d3.range(y.length))
      .join('rect')
      .attr('x', xCoord)
      .attr('y', yCoord)
      .attr('width', barWidth)
      .attr('height', barHeight)
      .attr('fill', (i) => zScale(z[i]));
  }, [
    updateAxis,
    y,
    isHorizontal,
    xScaleLinear,
    xScaleBand,
    x,
    crossScale,
    z,
    yScaleLinear,
    yScaleBand,
    zScale,
  ]);

  useEffect(() => {
    updateChart();
  });

  useUpdateEffect(() => {
    updateChart();
  }, [data, width, height]);

  return (
    <StyledSvg ref={svgRef} width={width} height={height} isHorizontal={!!isHorizontal}>
      <XAxis
        xScale={isHorizontal ? xScaleLinear : xScaleBand}
        removeDomain
        yOffset={height - chartMargin.bottom}
        yTickOffset={isHorizontal ? 31 : 24}
        ticks={4}
        tickSize={height - chartMargin.top - chartMargin.bottom}
        tickFormatFn={(d) => (formatXAxisTick ? formatXAxisTick(d) : `${d}`)}
        customCall={(el) => {
          el.selectAll('.tick text')
            .attr(
              'transform',
              (!isHorizontal && `translate(-${xScaleBand.bandwidth() / 2}, 0)`) || null
            )
            .style('text-anchor', isHorizontal ? '' : 'start');
          el.selectAll('.tick line').attr(
            'transform',
            !isHorizontal ? `translate(-${xScaleBand.bandwidth() / 2}, 0)` : null
          );
        }}
      />
      <YAxis
        yScale={isHorizontal ? yScaleBand : yScaleLinear}
        xOffset={chartMargin.left}
        removeDomain
        tickSize={width - chartMargin.left - chartMargin.right}
        tickFormatFn={(d) => _upperFirst(`${d}`)}
        tickValues={yTickValues}
        customCall={(el) => {
          const textItems = el.selectAll('.tick text');
          const textWidth = textItems
            .nodes()
            .map((item: d3.BaseType) => (item as Element).getBoundingClientRect().width);

          textItems.attr('x', (d, i) => (isHorizontal ? -45 : -textWidth[i] - 8)).attr('y', 1);

          el.selectAll('.tick line').attr(
            'transform',
            isHorizontal ? `translate(0, ${yScaleBand.bandwidth() / 2})` : null
          );
        }}
      />
      <g className="bars" />
    </StyledSvg>
  );
};

export default GroupedBarChart;
