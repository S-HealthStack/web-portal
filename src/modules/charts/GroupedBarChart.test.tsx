import React from 'react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components/';
import GroupedBarChart from 'src/modules/charts/GroupedBarChart';
import theme from 'src/styles/theme';

describe('GroupedBarChart', () => {
  it('should render', () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <GroupedBarChart
          data={[{ name: 'test', dataKey: 'test', value: 10 }]}
          width={1000}
          height={1000}
          barColors={['primary']}
          maxValue={0}
          numberOfKeys={0}
        />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('[NEGATIVE] should render with empty data', () => {
    const { baseElement, rerender } = render(
      <ThemeProvider theme={theme}>
        <GroupedBarChart
          data={[]}
          width={1000}
          height={1000}
          barColors={['primary']}
          maxValue={0}
          numberOfKeys={0}
        />
      </ThemeProvider>
    );

    expect(baseElement).toBeDefined();

    rerender(
      <ThemeProvider theme={theme}>
        <GroupedBarChart
          data={[{ name: '', dataKey: '', value: 0 }]}
          width={1000}
          height={1000}
          barColors={['primary']}
          maxValue={0}
          numberOfKeys={0}
        />
      </ThemeProvider>
    );
    expect(baseElement).toBeDefined();
  });
});
