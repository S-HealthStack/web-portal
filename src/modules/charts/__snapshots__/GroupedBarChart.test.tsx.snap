// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupedBarChart should render 1`] = `
.c1 > .tick > text {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
}

.c2 > .tick > text {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
}

.c0 .xAxis .tick line {
  stroke-opacity: 1;
  stroke: #F0F0F0;
}

.c0 .xAxis .tick:last-child line {
  stroke-dasharray: 1;
}

.c0 .xAxis .tick line {
  stroke-dasharray: 2;
}

.c0 .xAxis .tick text {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
}

.c0 .yAxis .tick line {
  stroke-opacity: 1;
  stroke: #F0F0F0;
}

.c0 .yAxis .tick:first-child line,
.c0 .yAxis .tick:last-child line {
  stroke-dasharray: 1;
}

.c0 .yAxis .tick line {
  stroke-dasharray: 2;
}

.c0 .yAxis .tick text {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
}

<body>
  <div>
    <svg
      class="c0"
      height="1000"
      width="1000"
    >
      <g
        class="c1 xAxis undefined"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="middle"
        transform="translate(0, 964)"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(520.5,0)"
        >
          <line
            stroke="currentColor"
            transform="translate(-467.5, 0)"
            y2="-964"
          />
          <text
            dy="0em"
            fill="currentColor"
            style="text-anchor: start;"
            transform="translate(-467.5, 0)"
            y="24"
          >
            test
          </text>
        </g>
      </g>
      <g
        class="c2 yAxis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="start"
        transform="translate(53,0)"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(0,482.5)"
        >
          <line
            stroke="currentColor"
            x2="935"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
            y="1"
          >
            0
          </text>
        </g>
      </g>
      <g
        class="bars"
      >
        <rect
          fill="#4475E3"
          height="0"
          width="0"
          x="520.5"
          y="482"
        />
      </g>
      <line
        class="rightChartBorder"
        style="stroke: #2D2D2D; stroke-width: 1; stroke-dasharray: 1; stroke-opacity: 0.2;"
        x1="988"
        x2="988"
        y1="0"
        y2="964"
      />
    </svg>
  </div>
</body>
`;
